tools/autograd/gen_python_functions.py

```diff
@@ -1349,7 +1351,28 @@ def go(f: NativeFunction) -> str:
         )
 
         if lambda_return == "void":
-            return f"""\
+            # Make in-place foreach return `self` at python-binding level.
+            # ref: https://github.com/pytorch/pytorch/pull/118622#pullrequestreview-1904804954
+            self_arg = f.func.arguments.self_arg
+            if (
+                str(f.func.name).startswith("_foreach_")
+                and f.func.kind() == SchemaKind.inplace
+                and (
+                    self_arg is not None and is_tensor_list_type(self_arg.argument.type)
+                )
+            ):
+                return f"""\
+{schema_comment}
+{inits}
+auto dispatch_{name} = []({lambda_formals}) -> ::std::vector<at::Tensor> {{
+  pybind11::gil_scoped_release no_gil;
+  {dispatch_callee}({dispatch_args});
+  return self.vec();
+}};
+return wrap(dispatch_{name}({lambda_args}){set_requires_grad});
```

Comment:
Not sure we actually want to return whatever is returned by dispatch here as this would construct a new tuple. We would just want to return the input as-is in this case as per Alban's comment.